File: backup/cyclesignal_flutter/lib/logic/phase.dart
import 'cycle.dart';

class CyclePosition {
  final int cycleDay;
  final int length;
  final String phase;
  final double moonPhase; // 0..1
  CyclePosition(this.cycleDay, this.length, this.phase, this.moonPhase);
}

// Current cycle-day + phase label for the dashboard hero (PRD 0.5).
CyclePosition currentPosition(List<Cycle> cycles) {
  if (cycles.isEmpty) return CyclePosition(1, 28, 'No data yet', 0);
  final current = cycles.last;
  final cycleDay = (diffDays(todayDate(), parseIso(current.startDate)) + 1).clamp(1, 999);
  final length = current.length;
  final ratio = (cycleDay / length).clamp(0.0, 1.0);

  String phase = 'Follicular phase';
  if (cycleDay <= 5) {
    phase = 'Menstrual phase';
  } else if (ratio > 0.45 && ratio < 0.6) {
    phase = 'Ovulatory phase';
  } else if (cycleDay > length - 10) {
    phase = 'Premenstrual phase';
  } else if (ratio >= 0.6) {
    phase = 'Luteal phase';
  }
  return CyclePosition(cycleDay, length, phase, ratio);
}

File: backup/cyclesignal_rn/app/src/lib/phase.ts
import { Cycle, diffDays, parseISO } from './cycle';

// Current cycle-day + phase label for the dashboard hero (PRD 0.5).
export function currentPosition(cycles: Cycle[]): {
  cycleDay: number;
  length: number;
  phase: string;
  moonPhase: number;
} {
  if (cycles.length === 0) {
    return { cycleDay: 1, length: 28, phase: 'No data yet', moonPhase: 0 };
  }
  const current = cycles[cycles.length - 1];
  const today = new Date(new Date().toISOString().slice(0, 10) + 'T00:00:00');
  const cycleDay = Math.max(1, diffDays(today, parseISO(current.startDate)) + 1);
  const length = current.length;
  const ratio = Math.min(1, cycleDay / length);

  let phase = 'Follicular phase';
  if (cycleDay <= 5) phase = 'Menstrual phase';
  else if (ratio > 0.45 && ratio < 0.6) phase = 'Ovulatory phase';
  else if (cycleDay > length - 10) phase = 'Premenstrual phase';
  else if (ratio >= 0.6) phase = 'Luteal phase';

  return { cycleDay, length, phase, moonPhase: ratio };
}
